pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/WindowFunction.java

```diff
@@ -18,14 +18,75 @@
  */
 package org.apache.pinot.query.runtime.operator.window;
 
+import com.google.common.collect.ImmutableMap;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
 import java.util.List;
+import java.util.Map;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.query.planner.logical.RexExpression;
+import org.apache.pinot.query.runtime.operator.WindowAggregateOperator;
+import org.apache.pinot.query.runtime.operator.utils.AggregationUtils;
+import org.apache.pinot.query.runtime.operator.window.aggregate.AggregateWindowFunction;
+import org.apache.pinot.query.runtime.operator.window.range.RangeWindowFunction;
+import org.apache.pinot.query.runtime.operator.window.value.ValueWindowFunction;
 
 
-public interface WindowFunction {
+/**
+ * This class provides the basic structure for window functions. It provides the batch row processing API:
+ * processRows(List<Object[]> rows) which processes a batch of rows at a time.
+ *
+ */
+public abstract class WindowFunction extends AggregationUtils.Accumulator {
```

Comment:
Suggest adding a `WindowFunctionFactory` if you want to make it pluggable, maybe following the same way of how `TransformFunction` is handle

Reply:
Also consider making this a separate PR?

Reply:
Added WindowFunctionFactory, will make a separated pr for pluggable stuffs